### PR TITLE
Plasma5: Use upstream xsession/wayland-session files

### PR DESCRIPTION
**Pull request by bkchr**
This also enables Wayland for plasma5.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by worldofpeace** [comment]
pinging #53843 because it will change that options name.

File: nixos/modules/services/x11/desktop-managers/plasma5.nix [hunk 1]
```diff
@@ -37,32 +37,6 @@ in
 
   config = mkMerge [
     (mkIf (xcfg.enable && cfg.enable) {
-      services.xserver.desktopManager.session = singleton {
-        name = "plasma5";
-        bgSupport = true;
-        start = ''
-          # Load PulseAudio module for routing support.
-          # See http://colin.guthr.ie/2009/10/so-how-does-the-kde-pulseaudio-support-work-anyway/
-          ${optionalString config.hardware.pulseaudio.enable ''
```
Comment (worldofpeace) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix]: Are all the extra things in here not needed?

If not you can maybe do something like https://github.com/NixOS/nixpkgs/blob/1a66e1dd276940e9666ff5fee1f1a739f3a55736/nixos/modules/services/x11/desktop-managers/pantheon.nix#L80

**Comment by worldofpeace** [comment]
I think the commit msg should start like `nixos/plasma5`

**Review by matthewbauer — APPROVED** [review]

File: nixos/modules/services/x11/desktop-managers/plasma5.nix (hunk 1 above)
Comment (matthewbauer) [reply nixos/modules/services/x11/desktop-managers/plasma5.nix]: Yeah that's a good idea actually. Those will still be run with upstream sessions?

Comment (bkchr) [reply nixos/modules/services/x11/desktop-managers/plasma5.nix]: I think we can drop this, if you look here: https://bugs.freedesktop.org/show_bug.cgi?id=95104

Comment (bkchr) [reply nixos/modules/services/x11/desktop-managers/plasma5.nix]: I tested switchting audio sources and worked without any problems for me.

**Comment by worldofpeace** [comment]
Testing in a vm shortly

**Comment by worldofpeace** [comment]
@bkchr I think it would be a good idea to possibly add a test for the wayland session?

Current test is broken because the session is now called `plasma`,
so we need to unset
```
services.xserver.desktopManager.default = "plasma5";
```
there.

And it seems that using sddm's autologin fails if we try to set the default session to `plasma`
```
- SDDM auto-login requires that services.xserver.desktopManager.default and
services.xserver.windowManager.default are set to valid values. The current
default session: plasma is not valid.
```

Not sure if the assertion is right there.

**Comment by bkchr** [comment]
Hmm okay :( I will take a look.

**Comment by matklad** [comment]
Did a quick test today by applying the patch on top of 19.03. 

Launching wayland session explicitly works (in fact, writing this comment from withing wayland session), but the default seession in SDDM becomes `xterm`, so you have to choose plasma explicitly.

**Comment by hedning** [comment]
> Not sure if the assertion is right there.

Right autologin with an upstream session doesn't work in sddm and lightdm at the moment. I left these asserts when doing https://github.com/NixOS/nixpkgs/pull/44497. I reasoned that the assert was in place to avoid getting stuck in a bad auto login (the upstream sessions isn't available at eval time). 

That's fixed in https://github.com/NixOS/nixpkgs/pull/53843, though it's a bit messy, partly due to the windowManager/desktopManager split which `sessionPackages` doesn't have. Though I think I've come up with a solution that would rework `desktopManager.default` so it could be used to pick sessions from `sessionPackages` instead of introducing a new `displayManager.defaultSession` option. That might be preferable.

**Comment by Infinisil** [comment]
As far as I can see, the only thing left to do is rename the commit to start with "nixos/plasma5". Anything else?

**Comment by worldofpeace** [comment]
@Infinisil  It works but there's the breakages with upstream sessions noted by @hedning, @matklad, and I.

I don't really think it's a good idea to just leave autologin broken so we need something like https://github.com/NixOS/nixpkgs/pull/53843

Other things are:

1. Fix the NixOS test (noted in https://github.com/NixOS/nixpkgs/pull/56988#issuecomment-470698604)
2. Possibly write a test for the wayland session
3. adjust commit msg (https://github.com/NixOS/nixpkgs/pull/56988#issuecomment-470361894 https://github.com/NixOS/nixpkgs/pull/56988#issuecomment-477691245)

**Comment by matthewbauer** [comment]
Rebased

**Comment by adisbladis** [comment]
The wayland session is failing with log (`~/.local/share/sddm/wayland-session.log`):
```
dbus-update-activation-environment: warning: error sending to systemd: org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.systemd1 was not provided by any .service files
qt.qpa.plugin: Could not load the Qt platform plugin "wayland-org.kde.kwin.qpa" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: wayland-org.kde.kwin.qpa, wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, minimal, offscreen, vnc, xcb.

/nix/store/9mlxan56vpqk33ask109pgqjwbiwl9xl-plasma-workspace-5.16.5/bin/startplasmacompositor: line 283:  2352 Aborted                 (core dumped) /nix/store/xvrlykra01fj12929nwf8sjpn7mlfd1c-kwin-5.16.5-bin/bin/kwin_wayland --xwayland --libinput --exit-with-session=/nix/store/9mlxan56vpqk33ask109pgqjwbiwl9xl-plasma-workspace-5.16.5/libexec/startplasma
startplasmacompositor: Shutting down...
startplasmacompositor: Done.
```

**Comment by worldofpeace** [comment]
@adisbladis It looks like something that `plasma-workspace-5.16.5/bin/startplasmacompositor` calls into isn't wrapped etc.

**Comment by Synthetica9** [comment]
Is this now ready for prime time?

**Comment by worldofpeace** [comment]
> Is this now ready for prime time?

See https://github.com/NixOS/nixpkgs/pull/56988#issuecomment-477703129.

To paraphrase:
1. Autologin is still broken in sddm if you use upstream sessions.
Needs https://github.com/NixOS/nixpkgs/pull/53843 or similar idea from @hedning 
2. there's no integration testing for plasma wayland
3. Something is improperly wrapped or missing a wrapper
https://github.com/NixOS/nixpkgs/pull/56988#issuecomment-539064285

Also, since the session names have changed, what will happen for users that have done
```
services.xserver.desktopManager.default = "plasma5";
```
for the xsession?

**Comment by solson** [comment]
> 1. Autologin is still broken in sddm if you use upstream sessions.
  Needs #53843 or similar idea from @hedning

Looks like the mentioned PR has since been merged.

**Comment by worldofpeace** [comment]
Correct, with https://github.com/NixOS/nixpkgs/pull/53843 merged you're not going to get any regressions with using upstream session. The options are renamed now as well.

So 2 and 3 now :smile:

**Comment by worldofpeace** [comment]
Moving this to 20.09 milestone. At a glance, it's not as simple as I thought. I need to know more about nixpkgs custom startup script for wayland.

**Comment by pasqui23** [comment]
I've "rebased"(copied from the source tree in my `/etc/nixos` then modified it) this pr and ,while the "plasma" session work perfectly(it even seem faster,probably by skipping all of our custom init) "plasmawayland" still crashes with:
```
giu 07 18:41:16 nixos xsession[28941]: qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
giu 07 18:41:16 nixos xsession[28941]: This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
giu 07 18:41:16 nixos xsession[28941]: Available platform plugins are: wayland-org.kde.kwin.qpa, eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, xcb.

```
in the journal.
Weird that a wayland session still look for xcb.

**Comment by pasqui23** [comment]
@bkchr maybe I can publish a pr to use only the upstream xsession files,and to enable the wayland sessions in a later pr,maybe waiting for https://github.com/NixOS/nixpkgs/pull/84542 ?

**Comment by hedning** [comment]
(Assuming you're using SDDM). @pasqui23 I'd try setting `SessionCommand` in the wayland section of sddm:
https://github.com/NixOS/nixpkgs/blob/148bf83ceeaabf862cb83cf1e988bc8761a054a4/nixos/modules/services/x11/display-managers/sddm.nix#L60-L63

This is done in X11 section:
https://github.com/NixOS/nixpkgs/blob/148bf83ceeaabf862cb83cf1e988bc8761a054a4/nixos/modules/services/x11/display-managers/sddm.nix#L53

**Comment by pasqui23** [comment]
I did apply @hedning's suggestion but unfortunately wayland still did not work.
Also my `.local/share/sddm/wayland-session.log` is empty.
To note that sddm's default `wayland-session` and `xsession` are wholly different.

https://github.com/sddm/sddm/blob/develop/data/scripts/Xsession
https://github.com/sddm/sddm/blob/develop/data/scripts/wayland-session 
Wayland's,after a  bit of initialization shared with xsession,jump straight to `exec "$@"`,while xsession first import a bunch of stuff from x-related files.

I don't know how that relates to wayland

**Comment by hedning** [comment]
Right, `wayland-session` looks like it does the correct things (similar to how GDM handles wayland sessions). Looks like it's most likely something that isn't wrapped correctly as noted in https://github.com/NixOS/nixpkgs/pull/56988#issuecomment-539075638.

**Comment by pasqui23** [comment]
However that still leaves .local/share/sddm/wayland-session.log empty even after my failed session,where else do I look?

**Comment by oxalica** [comment]
Upstream changed a lot. Is there anything in advance?

**Comment by pasqui23** [comment]
@oxalica I'm waiting for nixos-unstabe to proceed.

**Comment by pasqui23** [comment]
@bkchr I'm continuing your work in https://github.com/NixOS/nixpkgs/pull/100057
It's ok if you close this and continue the work on that?

**Comment by worldofpeace** [comment]
The last rebase was a long time ago so you are continuing the work @pasqui23. Closing in favor of that PR for now

**Comment by peterhoeg** [comment]
Guys, I'm spamming some of the old KDE wayland issues/PRs to ensure we get some testers. This comment is written from a plasma wayland session spawned from sddm after applying this PR (and the one-liner mentioned in a comment) which basically gets it going: https://github.com/NixOS/nixpkgs/pull/117102

**Pull request closed without merging**